V0.10.1

benjaminion opened the pull request:
Brings us up to the v0.10.0 core beacon chain spec. It includes a big BLS refactor to make the interfaces more uniform with the BLS signature standard.

_Note: does not include other v0.10.0/1 changes, such as fork choice_

Waiting for a couple of external things before we can merge
 - [x] Fixes to the BLS reference tests: https://github.com/ethereum/eth2.0-specs/pull/1575
 - [x] Fixes to the reference test deposit handling: https://github.com/ethereum/eth2.0-specs/issues/1582
 - [x] Fixes to the BLS reference tests: https://github.com/ethereum/eth2.0-specs/issues/1597

Also
 - [ ] need to update the MockStartDepositGeneratorTest input test data. I've disabled it for now.
 - [x] Fix the new benchmark code. I seem to have broken it.
 - [x] Make sure no other changes sneaked in in v0.10.1

We probably want to stay at 0.9.4 until a number of other clients are ready with 0.10.0

benjaminion committed:
Big BLS refactor

In preparation for spec 0.10.0.
WIP - temporarily breaks many things.

benjaminion committed:
Continued refactoring

benjaminion committed:
Move aggregation over to the new BLS

benjaminion committed:
Implements (and passes) v0.10.0 BLS reference tests

benjaminion committed:
Up to v0.10.0 except for some tests

benjaminion committed:
Passing Artemis unit tests

ajsutton commented on util/src/main/java/tech/pegasys/artemis/util/bls/BLSKeyPair.java:
```diff
@@ -28,20 +27,15 @@ public static BLSKeyPair random(int entropy) {
   public BLSSecretKey getSecretKey() {
-    return secretKey;
+    return new BLSSecretKey(keyPair.secretKey());
   }
```
It's generally better to avoid getters having to create the object they're returning.  Was there a reason we're now deferring the creation of these `BLSPublicKey` and `BLSSecretKey` objects?

Admittedly I normally don't like constructors creating new objects but given these are just wrappers the benefit of caching them seems to outweigh the slight unpleasantness of constructors doing "work".

benjaminion replied:
Yeah, I've reverted and improved this now. Not sure what I was thinking.

ajsutton commented on util/src/main/java/tech/pegasys/artemis/util/config/Constants.java:
```diff
@@ -210,21 +211,21 @@ public static void setConstants(String Constants) {
       BLS_WITHDRAWAL_PREFIX = Bytes.wrap(new byte[1]);
 
       // Time parameters
+      MIN_GENESIS_DELAY = 300;
```
Yay! Sanity!

ajsutton reviewed (commented):
I haven't tried to compare this with the actual BLS spec changes, but otherwise this all LGTM.

benjaminion committed:
Tidy up

In particular, be more assertive about encapsulating the Mikuli library.

benjaminion committed:
Add toString methods to keep benchmarks happy

benjaminion committed:
Update BLS tests to v0.10.1

Unfortunately, there remains a bug in the fastAggregateVerify reference test data,
so those will fail until the test suite is fixed.

benjaminion committed:
Spotless

benjaminion committed:
Remove G2Point from reference test code

benjaminion committed:
Modify build.gradle to get CircleCI to update reftest cache

ajsutton committed:
Merge branch 'master' of github.com:PegaSysEng/artemis into v0.10.0

# Conflicts:
#	ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java

ajsutton committed:
Merge branch 'master' of github.com:PegaSysEng/artemis into v0.10.0

ajsutton committed:
Merge branch 'master' of github.com:PegaSysEng/artemis into v0.10.0

# Conflicts:
#	ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java
#	ethereum/statetransition/src/test-support/java/tech/pegasys/artemis/statetransition/BeaconChainUtil.java
#	validator/client/src/main/java/tech/pegasys/artemis/validator/client/ValidatorClient.java

ajsutton committed:
Merge branch 'master' of github.com:PegaSysEng/artemis into v0.10.0

# Conflicts:
#	ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlockHeader.java
#	ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Fork.java

ajsutton committed:
Spotless.

ajsutton committed:
Adjust random seeds to ensure we actually get different blocks.

ajsutton committed:
Merge branch 'master' into v0.10.0

ajsutton committed:
Use new constant name.

ajsutton committed:
Fix one more reference.

ajsutton committed:
Merge branch 'master' of github.com:PegaSysEng/teku into v0.10.0

cemozerr committed:
Merge branch 'master' into v0.10.0

cemozerr committed:
Run spotless

cemozerr committed:
Upgrade to fork choice spec v0.10.1

cemozerr committed:
Run spotless

cemozerr committed:
Fix block condition checking order

cemozerr committed:
Run spotless

cemozerr committed:
compute_signing_root changes for Validator spec

cemozerr committed:
Fix signer

cemozerr committed:
Run spotless

cemozerr committed:
BroadcastAttestation event when correct block arrives

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java:
```diff
@@ -290,15 +285,21 @@ public static BlockImportResult on_block(
+        checkOnBlockPostStorageConditions(block, store);
+    if (maybeFailureOnPostStorageConditions.isPresent()) {
+      return maybeFailureOnPostStorageConditions.get();
+    }
```
The aim of these checks is to bail out before any processing is done if we don't have the block ancestor or other prerequisites.  We don't commit the transaction so any changes made to the store are discarded if we fail either of these checks.  So we should just combine them into one again.

cemozerr replied:
get_ancestor throws a NullPointerException if we combine these condition checks.

cemozerr replied:
Spec explicitly adds the block to the store before doing the 'postStorageConditionCheck' as I call here.

cemozerr replied:
Unless we add the block that is being processed here to the store, get_ancestor does not have access to this block and thus throws the NPE. In order to combine these checks and still not get the NPE, we'd need to change get_ancestor, which doesn't seem to be worth the trade-off especially since most of the fork choice is going to be modified during the protoArray change.

ajsutton replied:
Yeah, the spec has always added before doing all the checks and we deliberately moved away from that. :)

We already know that we have the block's parent on the chain, so we can just call `get_ancestor(store, block.getParent_root(), finalizedSlot);`.  Which is what we used to have.  This check hasn't actually changed, the spec just checks the slot first then the get_ancestor call whereas previously it did it the other way around.

cemozerr replied:
Alright reverting. That makes sense.

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java:
```diff
@@ -318,11 +319,13 @@ public static BlockImportResult on_block(
           store.setBestJustifiedCheckpoint(justifiedCheckpoint);
         }
         if (should_update_justified_checkpoint(store, justifiedCheckpoint)) {
+          // TODO: Fork Choice Spec does not necessarily ask us to store this checkpoint state
```
We expect that any justified checkpoint has a checkpoint state stored for it. Otherwise we'll throw `NullPointerException` when later processing attestations in `get_latest_attesting_balance`:
`BeaconState state = store.getCheckpointState(store.getJustifiedCheckpoint());`

The fork choice spec actually does this store because it says to run `on_attestation` on every received attestation whether by gossip included in a block.  Apart from the validation checks for attestations, that ensures that the target state is in `store.checkpoint_states` and that `store.latest_messages` gets updates based on the attestation.

So storing these checkpoint states is part of the `on_attestation` processing.  The other half is updating latest messages which a ticket is open for (but unclear on how important it is to do when we're syncing such old blocks).

Technically, we're storing fewer checkpoint states that the spec implementation.  `on_attestation` would store the checkpoint state for the target of any attestation, we only store it when there are enough attestations to make it the best justified or justified state.

cemozerr replied:
Makes sense.

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java:
```diff
@@ -335,27 +338,46 @@ public static BlockImportResult on_block(
     final Checkpoint finalizedCheckpoint = state.getFinalized_checkpoint();
     if (finalizedCheckpoint.getEpoch().compareTo(store.getFinalizedCheckpoint().getEpoch()) > 0) {
+      // TODO: Fork Choice Spec does not necessarily ask us to store this checkpoint state, so if we
+      // do error here and exit, we will possibly have acted differently than other nodes.
```
The only case where this actually fails is if we're asking `process_slots` to go backwards (ie get to a slot prior to the starting state slot) - since we're definitely not doing that, the exception handling is just annoying boilerplate.

Per the spec we should be applying all of `on_attestation` to all attestations in the block which would trigger this same processing anyway.

cemozerr replied:
I see.

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java:
```diff
@@ -335,27 +338,46 @@ public static BlockImportResult on_block(
+              > 0
+          || !get_ancestor(store, store.getJustifiedCheckpoint().getRoot(), finalized_slot)
+              .equals(store.getFinalizedCheckpoint().getRoot())) {
+        store.setJustifiedCheckpoint(state.getCurrent_justified_checkpoint());
```
We need to ensure the checkpoint state for this new justified checkpoint is stored.

cemozerr replied:
Done.

ajsutton commented on services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java:
```diff
@@ -381,6 +382,14 @@ public void setNodeSlotAccordingToDBStore(Store store) {
+      Bytes32 headBlockRoot = this.stateProcessor.processHead();
+      this.eventBus.post(new BroadcastAttestationEvent(headBlockRoot, nodeSlot));
+    }
+  }
```
This seems wrong.  We might import two blocks for the same slot and then this would cause us to produce two conflicting attestations and get slashed. We may also do that because this event is fired both here and in `process_slot` - if the `process_slot` fired first and then this block was imported we'd potentially wind up with two different targets for the produced attestations.

edit: I can see this is aiming to meet the first paragraph under https://github.com/ethereum/eth2.0-specs/blob/v0.10.1/specs/phase0/validator.md#attesting  which says we can publish attestations "early" if we already have the block for that slot.  I suspect we should have nice centralised logic for doing that coordination and not split it across multiple classes to be really sure we have it right.  Strongly suspect we should do it as a separate PR (and since it's not a consensus affecting change it can go straight to master).

cemozerr replied:
https://pegasys1.atlassian.net/browse/BC-309?atlOrigin=eyJpIjoiZDNmNGY0ZDk3ZmYzNDI2OWE0Yzk5NWRkMDdhNzY0MDMiLCJwIjoiaiJ9

cemozerr replied:
Reverted these changes in this PR.

ajsutton commented on validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java:
```diff
@@ -182,10 +185,15 @@ public void onBlockImported(ImportedBlockEvent event) {
+      UnsignedLong slot = event.getNodeSlot();
+      if (slot.compareTo(latestBroadcastAttestationEventSlot) <= 0) {
+        return;
+      }
```
I guess that resolves my concerns about getting slashed above, but this still feels very weird to me.

ajsutton commented on validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java:
```diff
@@ -212,6 +220,7 @@ public void onAttestationEvent(BroadcastAttestationEvent event) throws IllegalAr
       asyncProduceAttestations(
           attesterInformations, headState, getGenericAttestationData(headState, headBlock));
 
+      latestBroadcastAttestationEventSlot = slot;
```
I suspect we should do this in the opposite order.  Otherwise we risk publishing attestations and then failing unexpectedly before we update the `latestBroadcastAttestationEventSlot` which opens us to being slashed.

As mentioned above (sorry for scattering this across multiple comments) - let's pull this out of this PR and do it separately.

cemozerr replied:
I'd say let's keep this in this PR to be fully done with the v0.10.1 change and if we want to make it more robust we can make a ticket and come back to it later on.

cemozerr replied:
Unless this seems like it will definitely fail in production, after I put these lines in the opposite order, I see no problem merging this in for now.

ajsutton reviewed (commented):
Looks good.  Hopefully the comments around storing the checkpoint states will make that area clearer, and I suspect we should split the code to publish as soon as the node slot block is imported into a separate PR and make sure we have a nice clean way to do that.

cemozerr committed:
Resolve comments

cemozerr committed:
Merge remote-tracking branch 'remotes/origin/master' into benjaminion-v0.10.0

# Conflicts:
#	artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java
#	data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconChainHeadHandler.java
#	data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconChainHeadHandlerTest.java

cemozerr committed:
Fix issues relating to merge

cemozerr committed:
Fix tests

cemozerr committed:
Merge remote-tracking branch 'remotes/origin/master' into benjaminion-v0.10.0

# Conflicts:
#	services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java

cemozerr committed:
Fix merge

ajsutton committed:
Merge branch 'master' of github.com:PegaSysEng/teku into v0.10.0

cemozerr committed:
Revert block check changes

cemozerr committed:
Revert changes on broadcasting attestations

cemozerr committed:
Merge branch 'master' into v0.10.0

ajsutton commented on build.gradle:
```diff
@@ -209,7 +209,7 @@ allprojects {
 }
 
-def refTestVersion = 'v0.9.4'
+def refTestVersion = 'v0.10.1' // TODO: Remove this comment (just kicking the CircleCI cache)
```
We should probably remove that comment now. :)

cemozerr replied:
Done.

ajsutton reviewed (approved):
LGTM.

ajsutton committed:
Merge branch 'master' of github.com:PegaSysEng/teku into v0.10.0

cemozerr commented on build.gradle:
```diff
@@ -209,7 +209,7 @@ allprojects {
 }
 
-def refTestVersion = 'v0.9.4'
+def refTestVersion = 'v0.10.1' // TODO: Remove this comment (just kicking the CircleCI cache)
```
```suggestion
def refTestVersion = 'v0.10.1'
```

cemozerr committed:
Remove unnecessary comment
